lightning/src/util/persist.rs

```diff
@@ -249,10 +259,1015 @@ where
+						monitor_name.as_str(),
+						update_name.as_str(),
+						lazy,
+					)?;
+				}
+			}
+		}
+		Ok(())
+	}
+}
+
+impl<
+		ChannelSigner: WriteableEcdsaChannelSigner,
+		K: Deref,
+		L: Deref,
+		ES: Deref + Clone,
+		SP: Deref + Clone,
+	> Persist<ChannelSigner> for MonitorUpdatingPersister<K, L, ES, SP>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+	ES::Target: EntropySource + Sized,
+	SP::Target: SignerProvider + Sized,
+{
+	/// Persists a new channel. This means writing the entire monitor to the
+	/// provided [`KVStore`].
+	fn persist_new_channel(
+		&self, funding_txo: OutPoint, monitor: &ChannelMonitor<ChannelSigner>,
+		_monitor_update_call_id: MonitorUpdateId,
+	) -> chain::ChannelMonitorUpdateStatus {
+		// Determine the proper key for this monitor
+		let monitor_name = MonitorName::from(funding_txo);
+		let maybe_old_monitor = self.read_monitor(&monitor_name);
+		match maybe_old_monitor {
+			Ok((_, ref old_monitor)) => {
+				// Check that this key isn't already storing a monitor with a higher update_id (collision)
+				if old_monitor.get_latest_update_id() > monitor.get_latest_update_id() {
+					log_error!(
+						self.logger,
+						"Tried to write a monitor at the same outpoint {} with a higher update_id!",
+						monitor_name.as_str()
+					);
+					return chain::ChannelMonitorUpdateStatus::UnrecoverableError;
+				}
+			}
+			// This means the channel monitor is new.
+			Err(ref e) if e.kind() == io::ErrorKind::NotFound => {}
+			_ => return chain::ChannelMonitorUpdateStatus::UnrecoverableError,
+		}
+		// Serialize and write the new monitor
+		let mut monitor_bytes = Vec::with_capacity(
+			MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL.len() + monitor.serialized_length(),
+		);
+		monitor_bytes.extend_from_slice(MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL);
+		monitor.write(&mut monitor_bytes).unwrap();
+		match self.kv_store.write(
+			CHANNEL_MONITOR_PERSISTENCE_NAMESPACE,
+			CHANNEL_MONITOR_PERSISTENCE_SUB_NAMESPACE,
+			monitor_name.as_str(),
+			&monitor_bytes,
+		) {
+			Ok(_) => {
+				// Assess cleanup. Typically, we'll clean up only between the last two known full monitors.
+				if let Ok((_, old_monitor)) = maybe_old_monitor {
+					// Updates must be > the old monitor's update_id. If we can't add, the channel
+					// is already stored closed so we can skip this.
+					if let Some(start) = old_monitor.get_latest_update_id().checked_add(1) {
+						let end = if monitor.get_latest_update_id() == CLOSED_CHANNEL_UPDATE_ID {
+							// We don't want to clean the rest of u64, so just do possible pending updates. Note that
+							// we never write updates at `CLOSED_CHANNEL_UPDATE_ID`.
+							cmp::min(
+								start.saturating_add(self.maximum_pending_updates),
+								CLOSED_CHANNEL_UPDATE_ID - 1,
+							)
+						} else {
+							monitor.get_latest_update_id().saturating_sub(1)
+						};
+						// We should bother cleaning up only if there's at least one update expected.
```

Comment:
I don't quite understand this comment. Perhaps it could be re-worded?

Reply:
I revised both the comment and the code to be a little clearer. This is meant to debug-assert that we're deleting what we think are, and that we're not issuing deletes for things we know don't exist. The only time we don't know which updates is exist is when the channel is closed, since the update_id fast-forwards to u64::MAX.